RedundantModifer Should Violate Final Modifer on Unnamed Variables if jdkVersion GTE 22

child of #14942 

Check documentation: https://checkstyle.org/checks/modifier/redundantmodifier.html#RedundantModifier

---

```
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="RedundantModifier"/>
    <property name="jdkVersion" value="21"/>
  </module>
</module>
PS D:\CS\test> cat src/Test.java
public class Test {
    void m(Object o) throws Exception {
        try (final var a = lock()) {  // violation , resource is effective final
        }

        if (o instanceof R(final int _)) { // expected violation, unnamed pattern variable is effective final

        }

        final int _ = sideEffect(); // expected violation, unnamed variable is effective final
    }
    AutoCloseable lock() { return null; }
    record R(int x) { }
    int sideEffect() { return 0; }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml  src/Test.java                                           
Starting audit...
[ERROR] D:\CS\test\src\Test.java:3:14: Redundant 'final' modifier. [RedundantModifier]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 
```

---

[JLS](https://docs.oracle.com/javase/specs/jls/se22/html/jls-6.html#jls-6.3) :

> If a declaration does not include an identifier, but instead includes the keyword _ (underscore), then the entity cannot be referred to by name. The following kinds of entity may be declared using an underscore:

Like resources, unnamed variables are effectively final which means that whether the unnamed variable has final modifier or not. the variable can't be reassigned. I expect a violation on the redundant modifer check on unnamed variables.


---
**Describe what you want in detail.**

We should update `RedundantModifer`  to violate the `final` modifier on unnamed variables if `jdkVersion` is >= 22. If a user is on old Java versions (below Java 9). the final modifier will not be redundant because `_` is a normal identifier in this case. 

Example:

```
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="RedundantModifier"/>
    <property name="jdkVersion" value="8"/>
  </module>
</module>
PS D:\CS\test> cat src/Test.java
public class Test {
    void m(Object o) throws Exception {
        final int _ = sideEffect(); // ok, jdkVersion is 8. underscore can be used as normal identifier so final is not redundant
    }
    AutoCloseable lock() { return null; }
    record R(int x) { }
    int sideEffect() { return 0; }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml  src/Test.java                                           
Starting audit...
Audit done.
Checkstyle ends with 0 errors.

```

**Note**: This update also includes `final` modifier on unnamed lambda parameters and unnamed catch parameters.

--- a/config/pmd.xml
+++ b/config/pmd.xml
@@ -372,7 +372,8 @@
       | //ClassDeclaration[@SimpleName='SarifLogger']//MethodDeclaration[@Name='escape']
       | //ClassDeclaration[@SimpleName='LeftCurlyCheck'
         or @SimpleName='FinalLocalVariableCheck'
-        or @SimpleName='NPathComplexityCheck']
+        or @SimpleName='NPathComplexityCheck'
+        or @SimpleName='RedundantModifierCheck']
              //MethodDeclaration[@Name='visitToken']
       | //ClassDeclaration[@SimpleName='FallThroughCheck']
              //MethodDeclaration[@Name='isTerminated']

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheck.java
@@ -64,6 +64,9 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * {@code strictfp} modifier when using JDK 17 or later. See reason at
  * <a href="https://openjdk.org/jeps/306">JEP 306</a>
  * </li>
+ * <li>
+ * {@code final} modifier on unnamed variables when using JDK 22 or later.
+ * </li>
  * </ol>
  * <p>
  * interfaces by definition are abstract so the {@code abstract} modifier is redundant on them.
@@ -152,7 +155,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * Old JDK version numbering is supported (e.g. 1.8 for Java 8)
  * as well as just the major JDK version alone (e.g. 8) is supported.
  * Type is {@code java.lang.String}.
- * Default value is {@code "21"}.
+ * Default value is {@code "22"}.
  * </li>
  * <li>
  * Property {@code tokens} - tokens to check
@@ -178,7 +181,13 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ANNOTATION_DEF">
  * ANNOTATION_DEF</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
- * RECORD_DEF</a>.
+ * RECORD_DEF</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#PATTERN_VARIABLE_DEF">
+ * PATTERN_VARIABLE_DEF</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_CATCH">
+ * LITERAL_CATCH</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LAMBDA">
+ * LAMBDA</a>.
  * </li>
  * </ul>
  * <p>
@@ -214,9 +223,9 @@ public class RedundantModifierCheck
     };
 
     /**
-     *  Constant for jdk 21 version number.
+     *  Constant for jdk 22 version number.
      */
-    private static final int JDK_21 = 21;
+    private static final int JDK_22 = 22;
 
     /**
      *  Constant for jdk 17 version number.
@@ -230,7 +239,7 @@ public class RedundantModifierCheck
      * as well as just the major JDK version alone (e.g. 8) is supported.
      *
      */
-    private int jdkVersion = JDK_21;
+    private int jdkVersion = JDK_22;
 
     /**
      * Setter to set the JDK version that you are using.
@@ -275,6 +284,9 @@ public class RedundantModifierCheck
             TokenTypes.RESOURCE,
             TokenTypes.ANNOTATION_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.PATTERN_VARIABLE_DEF,
+            TokenTypes.LITERAL_CATCH,
+            TokenTypes.LAMBDA,
         };
     }
 
@@ -300,8 +312,17 @@ public class RedundantModifierCheck
             case TokenTypes.RECORD_DEF:
                 checkForRedundantModifier(ast, TokenTypes.FINAL, TokenTypes.LITERAL_STATIC);
                 break;
-            case TokenTypes.CLASS_DEF:
             case TokenTypes.VARIABLE_DEF:
+            case TokenTypes.PATTERN_VARIABLE_DEF:
+                checkUnnamedVariables(ast);
+                break;
+            case TokenTypes.LITERAL_CATCH:
+                checkUnnamedVariables(ast.findFirstToken(TokenTypes.PARAMETER_DEF));
+                break;
+            case TokenTypes.LAMBDA:
+                processLambdaParameters(ast);
+                break;
+            case TokenTypes.CLASS_DEF:
             case TokenTypes.ANNOTATION_FIELD_DEF:
                 break;
             default:
@@ -315,6 +336,44 @@ public class RedundantModifierCheck
         if (jdkVersion >= JDK_17) {
             checkForRedundantModifier(ast, TokenTypes.STRICTFP);
         }
+    }
+
+    /**
+     * Process lambda parameters.
+     *
+     * @param lambdaAst node of type {@link TokenTypes#LAMBDA}
+     */
+    private void processLambdaParameters(DetailAST lambdaAst) {
+        final DetailAST lambdaParameters = lambdaAst.findFirstToken(TokenTypes.PARAMETERS);
+        if (lambdaParameters != null) {
+            TokenUtil.forEachChild(lambdaParameters, TokenTypes.PARAMETER_DEF,
+                    this::checkUnnamedVariables);
+        }
+    }
+
+    /**
+     * Check if the variable is unnamed and has redundant final modifier.
+     *
+     * @param ast node of type {@link TokenTypes#VARIABLE_DEF}
+     *     or {@link TokenTypes#PATTERN_VARIABLE_DEF}
+     *     or {@link TokenTypes#PARAMETER_DEF}
+     */
+    private void checkUnnamedVariables(DetailAST ast) {
+        if (jdkVersion >= JDK_22 && isUnnamedVariable(ast)) {
+            checkForRedundantModifier(ast, TokenTypes.FINAL);
+        }
+    }
+
+    /**
+     * Check if the variable is unnamed.
+     *
+     * @param ast node of type {@link TokenTypes#VARIABLE_DEF}
+     *     or {@link TokenTypes#PATTERN_VARIABLE_DEF}
+     *     or {@link TokenTypes#PARAMETER_DEF}
+     * @return true if the variable is unnamed
+     */
+    private static boolean isUnnamedVariable(DetailAST ast) {
+        return "_".equals(ast.findFirstToken(TokenTypes.IDENT).getText());
     }
 
     /**

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/modifier/RedundantModifierCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/modifier/RedundantModifierCheck.xml
@@ -37,6 +37,9 @@
  {@code strictfp} modifier when using JDK 17 or later. See reason at
  &lt;a href="https://openjdk.org/jeps/306"&gt;JEP 306&lt;/a&gt;
  &lt;/li&gt;
+ &lt;li&gt;
+ {@code final} modifier on unnamed variables when using JDK 22 or later.
+ &lt;/li&gt;
  &lt;/ol&gt;
  &lt;p&gt;
  interfaces by definition are abstract so the {@code abstract} modifier is redundant on them.
@@ -120,12 +123,12 @@
  }
  &lt;/pre&gt;</description>
          <properties>
-            <property default-value="21" name="jdkVersion" type="java.lang.String">
+            <property default-value="22" name="jdkVersion" type="java.lang.String">
                <description>Set the JDK version that you are using.
  Old JDK version numbering is supported (e.g. 1.8 for Java 8)
  as well as just the major JDK version alone (e.g. 8) is supported.</description>
             </property>
-            <property default-value="METHOD_DEF,VARIABLE_DEF,ANNOTATION_FIELD_DEF,INTERFACE_DEF,CTOR_DEF,CLASS_DEF,ENUM_DEF,RESOURCE,ANNOTATION_DEF,RECORD_DEF"
+            <property default-value="METHOD_DEF,VARIABLE_DEF,ANNOTATION_FIELD_DEF,INTERFACE_DEF,CTOR_DEF,CLASS_DEF,ENUM_DEF,RESOURCE,ANNOTATION_DEF,RECORD_DEF,PATTERN_VARIABLE_DEF,LITERAL_CATCH,LAMBDA"
                       name="tokens"
                       type="java.lang.String[]"
                       validation-type="tokenSet">

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
@@ -154,6 +154,9 @@ public class RedundantModifierCheckTest
             TokenTypes.RESOURCE,
             TokenTypes.ANNOTATION_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.PATTERN_VARIABLE_DEF,
+            TokenTypes.LITERAL_CATCH,
+            TokenTypes.LAMBDA,
         };
         assertWithMessage("Invalid acceptable tokens")
             .that(actual)
@@ -420,6 +423,36 @@ public class RedundantModifierCheckTest
         };
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputRedundantModifierStrictfpWithDefaultVersion.java"),
+                expected);
+    }
+
+    @Test
+    public void testFinalUnnamedVariablesWithDefaultVersion() throws Exception {
+        final String[] expected = {
+            "18:26: " + getCheckMessage(MSG_KEY, "final"),
+            "24:9: " + getCheckMessage(MSG_KEY, "final"),
+            "34:18: " + getCheckMessage(MSG_KEY, "final"),
+            "44:14: " + getCheckMessage(MSG_KEY, "final"),
+            "51:14: " + getCheckMessage(MSG_KEY, "final"),
+            "54:18: " + getCheckMessage(MSG_KEY, "final"),
+            "65:53: " + getCheckMessage(MSG_KEY, "final"),
+            "69:53: " + getCheckMessage(MSG_KEY, "final"),
+            "69:70: " + getCheckMessage(MSG_KEY, "final"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputRedundantModifierFinalUnnamedVariables.java"),
+                expected);
+    }
+
+    @Test
+    public void testFinalUnnamedVariablesWithOldVersion() throws Exception {
+        final String[] expected = {
+            "40:14: " + getCheckMessage(MSG_KEY, "final"),
+            "47:14: " + getCheckMessage(MSG_KEY, "final"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                        "InputRedundantModifierFinalUnnamedVariablesWithOldVersion.java"),
                 expected);
     }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierFinalUnnamedVariables.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierFinalUnnamedVariables.java
@@ -1,0 +1,86 @@
+/*
+RedundantModifier
+tokens = (default)METHOD_DEF, VARIABLE_DEF, ANNOTATION_FIELD_DEF, INTERFACE_DEF, \
+         CTOR_DEF, CLASS_DEF, ENUM_DEF, RESOURCE, PATTERN_VARIABLE_DEF, LITERAL_CATCH, LAMBDA
+jdkVersion = (default)22
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.modifier.redundantmodifier;
+
+import java.util.function.BiFunction;
+
+public class InputRedundantModifierFinalUnnamedVariables {
+    void m(Object o) {
+        // violation below, 'Redundant 'final' modifier'
+        if (o instanceof final String _) { }
+        if (o instanceof final String __) { }
+        if (o instanceof final String s)
+        if (o instanceof final String _s) { }
+
+        // violation below, 'Redundant 'final' modifier'
+        final int _ = sideEffect();
+        final int __ = sideEffect();
+        final int x = sideEffect();
+        final int _x = sideEffect();
+
+    }
+
+    void m2(Object o) {
+        switch (o) {
+            // violation below, 'Redundant 'final' modifier'
+            case final String _ -> { }
+            case Float _ -> { }
+            case final Integer __ -> { }
+            case final Double s -> { }
+            default -> { }
+        }
+    }
+
+    void m3() {
+        // violation below, 'Redundant 'final' modifier'
+        try (final var a = lock();) {
+
+        } catch (final Exception e) {
+
+        }
+
+        // violation below, 'Redundant 'final' modifier'
+        try (final var _ = lock();) {
+
+        // violation below, 'Redundant 'final' modifier'
+        } catch (final Exception _) {
+
+        }
+    }
+
+    void m4() {
+        BiFunction<Integer, Integer, Integer> f = (final Integer a, final Integer b) -> {
+            return 5;
+        };
+
+        // violation below, 'Redundant 'final' modifier'
+        BiFunction<Integer, Integer, Integer> f2 = (final Integer _, final Integer b) -> {
+            return 5;
+        };
+
+        BiFunction<Integer, Integer, Integer> f3 = (final Integer _, final Integer _) -> {
+        // 2 violations above:
+        //                   'Redundant 'final' modifier'
+        //                   'Redundant 'final' modifier'
+            return 5;
+        };
+    }
+
+    int sideEffect() {
+        return 0;
+    }
+
+    AutoCloseable lock() {
+        return null;
+    }
+
+    record Point(int x, int y) { }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierFinalUnnamedVariablesWithOldVersion.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierFinalUnnamedVariablesWithOldVersion.java
@@ -1,0 +1,77 @@
+/*
+RedundantModifier
+tokens = (default)METHOD_DEF, VARIABLE_DEF, ANNOTATION_FIELD_DEF, INTERFACE_DEF, \
+         CTOR_DEF, CLASS_DEF, ENUM_DEF, RESOURCE, PATTERN_VARIABLE_DEF, LITERAL_CATCH, LAMBDA
+jdkVersion = 8
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.modifier.redundantmodifier;
+
+import java.util.function.BiFunction;
+
+public class InputRedundantModifierFinalUnnamedVariablesWithOldVersion {
+    void m(Object o) {
+        if (o instanceof final String _) { }
+        if (o instanceof final String __) { }
+        if (o instanceof final String s)
+        if (o instanceof final String _s) { }
+
+        final int _ = sideEffect();
+        final int __ = sideEffect();
+        final int x = sideEffect();
+        final int _x = sideEffect();
+    }
+
+    void m2(Object o) {
+        switch (o) {
+            case final String _ -> { }
+            case Float _ -> { }
+            case final Integer __ -> { }
+            case final Double s -> { }
+            default -> { }
+        }
+    }
+
+    void m3() {
+        // violation below, 'Redundant 'final' modifier'
+        try (final var a = lock();) {
+
+        } catch (final Exception e) {
+
+        }
+
+        // violation below, 'Redundant 'final' modifier'
+        try (final var _ = lock();) {
+
+        } catch (final Exception _) {
+
+        }
+    }
+
+    void m4() {
+        BiFunction<Integer, Integer, Integer> f = (final Integer a, final Integer b) -> {
+            return 5;
+        };
+
+        BiFunction<Integer, Integer, Integer> f2 = (final Integer _, final Integer b) -> {
+            return 5;
+        };
+
+        BiFunction<Integer, Integer, Integer> f3 = (final Integer _, final Integer _) -> {
+            return 5;
+        };
+    }
+
+    int sideEffect() {
+        return 0;
+    }
+
+    AutoCloseable lock() {
+        return null;
+    }
+
+    record Point(int x, int y) { }
+}

--- a/src/xdocs/checks/modifier/redundantmodifier.xml
+++ b/src/xdocs/checks/modifier/redundantmodifier.xml
@@ -38,6 +38,9 @@
             <code>strictfp</code> modifier when using JDK 17 or later. See reason at
             <a href="https://openjdk.org/jeps/306">JEP 306</a>
           </li>
+          <li>
+            <code>final</code> modifier on unnamed variables when using JDK 22 or later.
+          </li>
         </ol>
 
         <p>
@@ -161,7 +164,7 @@ public class ClassExtending extends ClassExample {
               <td>jdkVersion</td>
               <td>Set the JDK version that you are using. Old JDK version numbering is supported (e.g. 1.8 for Java 8) as well as just the major JDK version alone (e.g. 8) is supported.</td>
               <td><a href="../../property_types.html#String">String</a></td>
-              <td><code>&quot;21&quot;</code></td>
+              <td><code>&quot;22&quot;</code></td>
               <td>10.18.0</td>
             </tr>
             <tr>
@@ -188,6 +191,12 @@ public class ClassExtending extends ClassExample {
                     ANNOTATION_DEF</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
                     RECORD_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#PATTERN_VARIABLE_DEF">
+                    PATTERN_VARIABLE_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_CATCH">
+                    LITERAL_CATCH</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LAMBDA">
+                    LAMBDA</a>
                   .
               </td>
               <td>
@@ -211,6 +220,12 @@ public class ClassExtending extends ClassExample {
                     ANNOTATION_DEF</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
                     RECORD_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#PATTERN_VARIABLE_DEF">
+                    PATTERN_VARIABLE_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_CATCH">
+                    LITERAL_CATCH</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LAMBDA">
+                    LAMBDA</a>
                   .
               </td>
               <td>3.0</td>

--- a/src/xdocs/checks/modifier/redundantmodifier.xml.template
+++ b/src/xdocs/checks/modifier/redundantmodifier.xml.template
@@ -38,6 +38,9 @@
             <code>strictfp</code> modifier when using JDK 17 or later. See reason at
             <a href="https://openjdk.org/jeps/306">JEP 306</a>
           </li>
+          <li>
+            <code>final</code> modifier on unnamed variables when using JDK 22 or later.
+          </li>
         </ol>
 
         <p>

